src/ansiblelint/rules/VariableNamingRule.py

```diff
@@ -0,0 +1,109 @@
+from typing import TYPE_CHECKING, List, Any, Dict, Union
+if TYPE_CHECKING:
+    from ansiblelint.constants import odict
+from ansiblelint.errors import MatchError
+from ansiblelint.file_utils import Lintable
+from ansiblelint.rules import AnsibleLintRule
+from ansiblelint.utils import parse_yaml_from_file
+import ansible.parsing.yaml.objects
+
+import re
+
+
+# properties/parameters are prefixed and postfixed with `__`
+def is_property(k):
+    return (k.startswith('__') and k.endswith('__'))
+
+
+def is_invalid_variable_name(text):
+    patterns = '^[a-z0-9_]*$'
+    if re.search(patterns, text):
+        return False  # string uses acceptable characters
+    else:
+        return True  # string uses unacceptable characters
+
+
+class VariableNamingRule(AnsibleLintRule):
+    id = 'var-naming'
+    base_msg = 'All variables should be named using only lowercase and underscores'
+    shortdesc = base_msg
+    description = 'All variables should be named using only lowercase and underscores'
+    severity = 'MEDIUM'  # ansible-lint displays severity when with --parseable-severity option
+    tags = ['formatting', 'readability']
```

Comment:
Add 'experimental' to this list, we have a rule of adding new rules as experimental, which makes them only
warnings (not affecting final outcome). We remove experimental on a major release, once we have confirmation that they prove not to create havoc.